NetKet/Hilbert/hilbert.hpp

```diff
@@ -17,65 +17,69 @@
 
 #include <memory>
 #include <set>
+#include "Graph/graph.hpp"
 #include "Utils/json_utils.hpp"
 #include "Utils/parallel_utils.hpp"
+#include "Utils/python_helper.hpp"
 #include "abstract_hilbert.hpp"
 #include "bosons.hpp"
 #include "custom_hilbert.hpp"
-#include "next_variation.hpp"
 #include "qubits.hpp"
 #include "spins.hpp"
 
 namespace netket {
 
 class Hilbert : public AbstractHilbert {
-  std::shared_ptr<AbstractHilbert> h_;
+  std::unique_ptr<AbstractHilbert> h_;
 
  public:
-  explicit Hilbert() {}
-
-  explicit Hilbert(const Hilbert &oh) : h_(oh.h_) {}
-
-  explicit Hilbert(const json &pars) { Init(pars); }
+  template <class Ptype>
+  explicit Hilbert(const Ptype &pars) {
+    InitWithoutGraph(ParsConv(pars));
+  }
 
-  void Init(const json &pars) {
-    CheckInput(pars);
+  template <class Ptype>
+  explicit Hilbert(const Graph &graph, const Ptype &pars) {
+    InitWithGraph(graph, ParsConv(pars));
+  }
 
-    if (FieldExists(pars["Hilbert"], "Name")) {
-      if (pars["Hilbert"]["Name"] == "Spin") {
-        h_ = std::make_shared<Spin>(pars);
-      } else if (pars["Hilbert"]["Name"] == "Boson") {
-        h_ = std::make_shared<Boson>(pars);
-      } else if (pars["Hilbert"]["Name"] == "Qubit") {
-        h_ = std::make_shared<Qubit>(pars);
-      }
-    } else {
-      h_ = std::make_shared<CustomHilbert>(pars);
-    }
+  json ParsConv(const json &pars) {
+    CheckFieldExists(pars, "Hilbert");
+    return pars["Hilbert"];
   }
+  pybind11::kwargs ParsConv(const pybind11::kwargs &pars) { return pars; }
 
-  void CheckInput(const json &pars)
-  {
-    if (!FieldExists(pars, "Hilbert")) {
-      if (!FieldExists(pars, "Hamiltonian")) {
-        throw InvalidInputError("Not enough information to construct Hilbert space");
-      } else {
-        if (!FieldExists(pars["Hamiltonian"], "Name")) {
-          throw InvalidInputError("Not enough information to construct Hilbert space");
-        }
-      }
-    }
+  template <class Ptype>
+  void InitWithoutGraph(const Ptype &pars) {
+    int size = FieldVal<int>(pars, "Size");
 
-    if (FieldExists(pars["Hilbert"], "Name")) {
-      std::set<std::string> hilberts = {"Spin", "Boson", "Qubit"};
+    json gpars;
+    gpars["Graph"]["Name"] = "Custom";
+    gpars["Graph"]["Size"] = size;
+    Graph graph(gpars);
```

Comment:
Am I missing something or does this result in a memory leak? Concrete classes inheriting from `abstract_hilbert` usually store a reference to the given `Graph`. This reference becomes dangling when `graph` goes out of scope.

Reply:
Yes, this is a problem, and it is again related to the complex constructor logic we want to get rid of (see also point 1 in the comment I just posted). We can get rid of this using free `MakeGraphAndHilbert` functions or similar